docs/decisions/2025-03-31-streaming-responses-for-chatbot.md

```diff
@@ -0,0 +1,166 @@
+
+## Decision Drivers
+
+* User experience (responsiveness, feedback, and engagement)
+* Reliabile connection to the chatbot
+* Implementation effort for both Nava and third-party clients
+* Maintainability
+
+## Considered Options
+
+### Polling
+
+- Handles intermittent connectivity through retry logic
+- Introduces latency dependent on polling intervals  
+- Requires writing logic for managing accumulated response chunks and state.
+
+### Server-Sent Events (SSE)
+
+- Immediate/event-driven updates provide a responsive user experience
+- Built-in browser support simplifies client-side implementation 
+- Straightforward server-side implementation
+
+### WebSockets
+
+- Supports two-way real-time communication
+- Flexible for interactive applications
+- More complex to implement and manage.  
+
+## Decision Outcome
+
+We chose Server-Sent Events (SSE) because it provides an event-driven approach to streaming responses with relatively low implementation complexity. SSE allows the server to push partial responses to the client as they become available, providing visual feedback that is engaging to users.
+
+Polling achieves similar UX results with periodic fetching accumulated response chunks from the server. But this approach introduces latency dependent on polling intervals and requires additional logic to manage state and retries. While LiteLLM supports asynchronous streaming, polling would require managing accumulated chunks and intervals.
+
+WebSockets offer two-way communication, good for real-time interactive applications. But WebSockets introduce more components to maintain for our current use case, requiring explicit connection management and reconnection handling.
+
+### Positive Consequences
+
+* Real-time streaming of partial responses increases user engagement
+* Users get immediate feedback that the system is working, rather than waiting for a complete response
+* LiteLLM and FastAPI have existing support for streaming
+* Low implementation effort
+
+### Negative Consequences
+
+- Requires client-side updates to support SSE connections (same for other all options).
+- SSE connections are limited to one-way communication (server-to-client only).
+
+## Requirements
+
+### Server-Side
+
+- Add an SSE endpoint (`/query_sse`) using FastAPI's `EventSourceResponse`.
+- Modify existing LLM query logic to yield response chunks as they become available.
+- Standardize event structure (`message`, `done`, `error`) for client-side handling.
+- Implement robust error handling to gracefully manage exceptions.
+
+### Client-Side
+
+- Initiate SSE connection using browser's `EventSource` API.
+- Handle incoming events (`message`, `done`, `error`) and progressively render partial responses.
+- Manage connection lifecycle, including error handling and reconnection logic.
```

Comment:
😕 Line 49 seems to suggest that we didn't need to do "explicit connection management and reconnection handling" for SSE